### PR TITLE
Add answer feature

### LINKED ISSUE #58: Adding answers to question feature
The feature will allow us to comment on questions.

Tasks:
1. Add text editor to display-question page  (#44 )
2. Create POST request when submitting, and save the comment in Answers table in the db.
3. Change the text editor in admin page.

### PR DESCRIPTION
**Pull request by idokk1**
Enable to add answer with ckeditor.
Included tests for adding new answer .

fix #58

### CODE REVIEW

**Comment by idokk1** [comment]
@ifireball do you think I need to move the create profile test to another PR or it's fine here ?

**Comment by ifireball** [comment]
> @ifireball do you think I need to move the create profile test to another PR or it's fine here ?

What do you keep talking about the test, as if you do not also include the code being tested? The important part is the code, not the test... 

In general PRs should include one thing, they can sometimes include more then one thinng if they are small enough.

File: home/tests.py
Comment (ifireball) [inline home/tests.py:330]: This is a very strange testing technique, its usually better to test views via the client and not try to call then directly as functions. 
Because you're trying to call the view directly you need to create a whole bunch of stuff to pass to if such as that post object that you're creating by crating the form object... this test is actually leaking a whole bunch of internal implementation details....

Wouldn't it be simpler to have the client perform an HTTP post to the view and then inspect the response? See how your team mates did similar thing in their PRs.

**Review by ifireball — CHANGES_REQUESTED** [review]

**Comment by idokk1** [comment]
@ifireball fixed what you asked.
In addition I added some tests on the html - if form is hidden for visitor and visible for authenticated user ,
if user is not authenticated if the login button is visible for him.
And I expanded the existing tests. 
if you can review please.

**Comment by rebeccaTubman** [comment]
LGTM
Great idea moving the `question_test_data` outside of the class to re-use it.

File: home/tests.py
Comment (avivz450) [inline home/tests.py:97]: Great job,
After PR #65 will be merged you can replace these lines with `return request.profile`

File: home/views.py
Comment (avivz450) [inline home/views.py:28]: Great job,
After PR #65  will be merged you can replace these lines with return `addAnswer(form, request.profile, question)`

**Review by avivz450 — APPROVED** [review]

**Comment by danitLevi** [comment]
LGTM

**Comment by danitLevi** [comment]
LGTM

**Comment by idokk1** [comment]
@ifireball this PR ready for review
I know this is a big PR, but i don't have an idea how to divide it..
its contains only the answers form with ckeditor, tests for that.

File: home/tests/test_add_answer.py
Comment (ifireball) [inline home/tests/test_add_answer.py:63]: why are the test functions below inside another test function?

File: home/tests/test_question_tag_model_functions.py
Comment (ifireball) [inline home/tests/test_question_tag_model_functions.py:7]: why is this soperfecial code move needed as part of this PR?

Comment (idokk1) [reply home/tests/test_question_tag_model_functions.py:7]: I wanted to use this fixture for my tests, so i moved it out of this class instead of copy the entire function.

**Comment by ifireball** [comment]
WRT to the size of the PR - I see 3 separate elements here:
1. The new answer form (I think its about answers but I'm not sure because the variables that refer to them and not always named consistently)
2. The core functionality for creating profiles (I suppose that after this gets in you'll need to refactor alot of old code to use it)
3. The chages to the questions page (some of those look like superficial renaming and moving of thing)

So the mix of those makes it harder to review the Pr and figure out what is relevant. But since its been going for so ling, I suppose  we can keep it like it is. But I did found one thing that looks like a bug in the testing code.

File: home/tests/test_question_tag_model_functions.py
Comment (ifireball) [reply home/tests/test_question_tag_model_functions.py:7]: Oh, you don't share fixture like that - to share fixtures you put then in a `conftest.py` file (look it up in the pytest docs).

But in any case you could have written that answer in the PR description and then I wouldn't have had to ask...

**Comment by idokk1** [comment]
create conftest.py and move the creating profile methods to another PR #113

File: home/tests/test_add_answer.py
Comment (idokk1) [reply home/tests/test_add_answer.py:63]: `test_default_params_answer` 
`test_get_answers`
you are talking about this functions ?

I'm not sure i understand what do you mean.

Comment (ifireball) [reply home/tests/test_add_answer.py:63]: The are indented. So they are inside `test_html_not_authenticated_user_form_visible`. look at your test results, they are probably not running as individual tests.

Comment (idokk1) [reply home/tests/test_add_answer.py:63]: oh, that was by mistake, I will fix that in the next commit when #113  will merge

**Comment by idokk1** [comment]
@ifireball the problems you told me about resolved.

File: home/tests/test_add_answer.py
Comment (liornoy) [inline home/tests/test_add_answer.py]: maybe use the authenticated_user fixture that's in conftest.py?

File: home/tests/test_add_answer.py
Comment (liornoy) [inline home/tests/test_add_answer.py]: what profile.create has to do with testing add answer feature

File: home/tests/test_add_answer.py
Comment (liornoy) [inline home/tests/test_add_answer.py]: question_test_data is a fixture defined in confest.py now not in test_question_tag_model_functions , so maybe this import is not relevant anymore

File: home/tests/test_question_tag_model_functions.py
Comment (liornoy) [inline home/tests/test_question_tag_model_functions.py]: need to remove this as it is now in conftest.py

File: Pipfile.lock
Comment (liornoy) [inline Pipfile.lock:70]: ckeditor is already merged to the code of master, so maybe rebase

File: home/tests/test_add_answer.py
Comment (liornoy) [inline home/tests/test_add_answer.py]: Why didn't you rebased it with the previous commit? I just spent time reviewing a commit that's apparently is not relevant because u made changes in it...

Comment (idokk1) [reply home/tests/test_add_answer.py]: I did fixup instead of squash, sorry for that my mistake.. 
fixed it now

**Review by rebeccaTubman — COMMENTED** [review]
I see you resolved the conflicts. 
Good job! I see test for the model and view functions including redirection. 
I did notice one minor thing - I don't see minimum value for answer content, does that mean you can add a black answer?

**Comment by idokk1** [comment]
> I see you resolved the conflicts.
> Good job! I see test for the model and view functions including redirection.
> I did notice one minor thing - I don't see minimum value for answer content, does that mean you can add a black answer?

for now you do

**Comment by rebeccaTubman** [comment]
> > I see you resolved the conflicts.
> > Good job! I see test for the model and view functions including redirection.
> > I did notice one minor thing - I don't see minimum value for answer content, does that mean you can add a black answer?
> 
> for now you do

OK :)

**Review by rebeccaTubman — APPROVED** [review]

**Review by danitLevi — APPROVED** [review]

**Comment by idokk1** [comment]
@ifireball the PR is ready for review. Thanks

File: home/tests/test_add_answer.py
Comment (ifireball) [inline home/tests/test_add_answer.py:36]: Missing test for unauthenticated user trying to post

File: home/views.py
Comment (ifireball) [inline home/views.py:15]: Tthis is not a Pythonic function name

Also why not define the args with explicit names instead of parsing to opaque `kwargs` structure?

File: home/views.py
Comment (ifireball) [inline home/views.py:18]: Can use the dict's `get` method to make this simpler instead of an `if` expression

File: home/views.py
Comment (ifireball) [inline home/views.py:28]: not sure moving the code out to a function makes sense, and the risk is that people migt think that helper function is a view function since its in the views file.

The function name is also not Pythonic.

**Review by ifireball — CHANGES_REQUESTED** [review]

File: home/views.py
Comment (idokk1) [reply home/views.py:15]: I added some functions to this function, because my feature is in the same view as Lior's feature.
the use of kwargs is was on his PR

File: home/views.py
Comment (idokk1) [reply home/views.py:18]: same here, this is content of Lior's PR.

File: home/views.py
Comment (ifireball) [reply home/views.py:15]: 1. Blaming your friend is not nice.
2. If its broken - fix it.

File: home/views.py
Comment (idokk1) [reply home/views.py]: fixed

File: home/tests/test_add_answer.py
Comment (idokk1) [reply home/tests/test_add_answer.py:37]: added

File: home/views.py
Comment (idokk1) [reply home/views.py]: not blaming him, just afraid to touch his code (that already merged) 
but I'll do it

Comment (ifireball) [reply home/views.py]: But you did touch it... Also why fear, you have tests no? Don't you trust the tests to tell you if you've broken anything important?

File: home/views.py
Comment (idokk1) [reply home/views.py]: fixed

File: home/views.py
Comment (idokk1) [reply home/views.py]: you right, fixed.

**Comment by idokk1** [comment]
@ifireball I fixed the last reviews and resolved the conflicts.

File: home/tests/test_add_answer.py
Comment (ifireball) [inline home/tests/test_add_answer.py:54]: I'm pretty sure this is supposed to return some kind of an HTTP error code and possibly display some error page, not raise an exception.

File: home/views.py
Comment (ifireball) [inline home/views.py:16]: what is the point of having a nested function like this, it seems to be just adding some unnecessary complexity to the code...

Comment (idokk1) [reply home/views.py]: removed

File: home/tests/test_add_answer.py
Comment (idokk1) [reply home/tests/test_add_answer.py:54]: but unauthenticated cannot see the form to add answer at all (and I checked it).
and if someone try to post to this page somehow without being authenticated user it will fail.
it's not enough?

Comment (ifireball) [reply home/tests/test_add_answer.py:54]: "seeing" the form is not required for doing an HTTP POST request - remember that the browser and the server are connected via an untrusted network.

You need to check that your site is generating the failure as needed, otherwise you have a security issue.

This test is using the client to do an HTTP request to the site. When you are doing an HTTP request to a site you are supposed to get some kind of an HTTP response (Errors are valid responses too). If you get an exception from the client it means it somehow failed to communicate.

Also, are you sure you need all the fixtures for this test?

Comment (idokk1) [reply home/tests/test_add_answer.py]: added a check that if there is no user , raise 404. and created test for that.

**Review by rebeccaTubman — APPROVED** [review]

File: home/views.py
Comment (liornoy) [inline home/views.py:32]: Maybe switch it to 401 response - unauthorized. (Barak recommended it to me also in a similar case)
you can see how I did it in my PR in views.py

File: home/tests/test_add_answer.py
Comment (liornoy) [inline home/tests/test_add_answer.py:54]: Maybe switch response to unauthorisde, see my comment below

File: home/views.py
Comment (idokk1) [reply home/views.py]: you right. 404 is really doesn't the right status code for this case

File: home/templates/home/question_detail.html
Comment (ifireball) [inline home/templates/home/question_detail.html:47]: This may or may not be a security issue - does the RichTextField class perfom any HTML sanitation to ensure there are not things like `<script>` tags?

**Review by ifireball — CHANGES_REQUESTED** [review]

File: home/templates/home/question_detail.html
Comment (idokk1) [reply home/templates/home/question_detail.html:47]: @ifireball fixed it in #131 , wait for merge and then i will use the same solution here.

Comment (idokk1) [reply home/templates/home/question_detail.html:47]: fixed

**Comment by idokk1** [comment]
@ifireball , fixed the security issue and resolved the rest of the reviews comments.

File: home/forms.py
Comment (ifireball) [inline home/forms.py:26]: Should this be the `RichTextBleackFeild` too? Also id expect to see `| safe` somewhere as a result of this change.

Comment (idokk1) [reply home/forms.py]: I changed it, 
the important thing is that `RichTextBleackField ` will be on the model , because there I preform the validation. 
the  `RichTextBleackField ` is same as `RichTextField` but with the validations.
and there is `|safe` on the question_detail.html ( `<p>{{ answer.content | safe }}</p>` )

**Comment by idokk1** [comment]
the PR is ready for re-review @ifireball

**Review by ifireball — APPROVED** [review]

**Pull request merged**
